### PR TITLE
Fixes #25990 - no reverse DNS search for bootserver

### PR DESCRIPTION
**Pull request by lzap**
This moves bootserver DNS PTR resolution from orchestration and NIC
classes to proxy API. This allows reporting TFTP servername correctly in
Proxy status pages and API.

@xprazak2

### CODE REVIEW

**Comment by ekohl** [comment]
This does look like an effective change in behavior. If the response from the proxy is a DNS name it's no longer converted into an IP. This might be a good thing, but was this by design?

File: lib/proxy_api/tftp.rb
Comment (timogoebel) [inline lib/proxy_api/tftp.rb]: This is very unclean. The Rendering scopes should only be used for template rendering. Can you please move dns resolotion to a dedicated service?

**Review by timogoebel — DISMISSED** [review]

**Comment by lzap** [comment]
> This does look like an effective change in behavior. If the response from the proxy is a DNS name it's no longer converted into an IP. This might be a good thing, but was this by design?

Don't understand, I still aim to convert it to IP. It must be an IP because ISC DHCP won't allow to create a record with hostname for next-server. Where exactly do you see the change?

File: lib/proxy_api/tftp.rb
Comment (lzap) [reply lib/proxy_api/tftp.rb]: Yeah I could not figure it out, so like new module and including it in both places?

**Comment by lzap** [comment]
So here's the thing. I rebase to the branch and seed/migrate but when I hit Refresh Features Foreman reaches out to `/features` and I'd expect it to use the new endpoint. No V2 request, just reaching out to `/features/` directly. 

What I am doing wrong?

File: lib/proxy_api/tftp.rb
Comment (ares) [reply lib/proxy_api/tftp.rb]: I think a service class would be better then module. Something like 
```
class DnsResolver
  def self.dns_lookup(name_or_ip)
    resolver = Resolv::DNS.new
     Timeout.timeout(Setting[:dns_conflict_timeout]) do
      begin
        resolver.getname(name_or_ip)
      rescue Resolv::ResolvError
        resolver.getaddress(name_or_ip)
      end
    end
  rescue StandardError => e
    log_warn "Template helper dns_lookup failed: #{e}"
    raise e
  end
end
```

and then here

```
DnsResolver.dns_lookup(URI.parse(@url).host)
```

and in dns_lookup macro

```
DnsResolver.dns_lookup(name_or_ip)
```

**Comment by ares** [comment]
> This does look like an effective change in behavior. If the response from the proxy is a DNS name it's no longer converted into an IP. This might be a good thing, but was this by design?

perhaps there was a misunderstanding, the dns_lookup method (despite its name) converts IP -> name but also name -> IP, based on what the argument is. Or I misunderstood completely :-)

**Comment by lzap** [comment]
Can someone take a quick look at https://github.com/theforeman/foreman/pull/6861 before I rebase this one? It's relevant and it will cause conflicts if I start working on both.

File: lib/proxy_api/tftp.rb
Comment (ekohl) [inline lib/proxy_api/tftp.rb]: This can actually also be a FQDN. Previously this would be resolved but the current implementation would break.

File: test/unit/proxy_statuses/proxy_status_tftp_test.rb
Comment (ekohl) [inline test/unit/proxy_statuses/proxy_status_tftp_test.rb]: Could you add a test where it returns a FQDN?

**Review by ekohl — CHANGES_REQUESTED** [review]

File: lib/proxy_api/tftp.rb
Comment (lzap) [reply lib/proxy_api/tftp.rb]: Oh I see, I thought that installer only accepts IPv4 for some reason. Will fix this.

Comment (ekohl) [reply lib/proxy_api/tftp.rb]: If only. I actually wouldn't mind deprecating that and adding a warning in the proxy when it sees a non-IP. Then on the Foreman side we would change the behavior to trust the response if it's an IP rather than doing the ip -> name -> ip dance.

File: app/models/concerns/orchestration/dhcp.rb
Comment (lzap) [inline app/models/concerns/orchestration/dhcp.rb:91]: One more thing before I continue - there is this helper class called `NicIpResolver` which performs one quite confusing thing IMHO. It tries to query nameservers associated with the NIC-Subnet and if that fails it tries OS-level query via Ruby. This can be confusing when Subnet is misconfigured or entry is missing on the managed DNS server. Two options: A) I keep this behavior. B) I only will use OS-level.

Option (A) is more tricky because I am actually moving this out of our model context where NIC and Subnet is available to Proxy APi context where it's used from two places.

That's not all, there is also setting named "query_local_nameservers" and when set Foreman is supposed to use OS-level resolver instead reaching out to managed ones. However this code which I am trying to refactor completely ignores that setting.

If I want to do (B) or even (B) + utilizing the setting, then it's a bit off a problem when this is called from the ProxyStatus because there is no host context, therefore Subnet (thus managed DNS servers) is not known. If we want to fix the original bug report (status page) then I think option (A) is the only one but this creates confusion around the setting.

@ekohl and @ares

Comment (lzap) [reply app/models/concerns/orchestration/dhcp.rb:91]: What a mess! So the `NicIpResolver` actually:

* tries to resolve against authorized DNS servers defined in the domain
* then tries to resolve against:
 * against authorized DNS servers defined in the domain when query_local_nameservers is false
 * against OS DNS when query_local_nameservers is true

I am starting to think this needs a deeper redesign, probably we can introduce new Subnet flag TFTP server IPv4 address stored in Foreman database and some migration that will help users to fill it (proxy servers would need to be up during the upgrade or DNS correctly set). WHAT A MESS.

Comment (ekohl) [reply app/models/concerns/orchestration/dhcp.rb:91]: I wonder if we should simplify this. Reading https://tools.ietf.org/html/rfc2132#section-9.4 nothing indicates this needs to be an IP but can actually be a DNS name. The ISC DHCP manual also suggests it's a text rather than an IP, indicating to me it can use DNS. So why do we need to do this translation at all? Perhaps this is a good example of garbage in, garbage out.

I'd appreciate it if you could try this out.

**Comment by lzap** [comment]
Rebased a WIP version to see how many tests fail, but we need to discuss the above. This is really bad code.

File: app/models/concerns/orchestration/dhcp.rb
Comment (lzap) [reply app/models/concerns/orchestration/dhcp.rb:91]: From ISC DHCP man page: _The next-server statement is used to specify the host address of the server  from  which  the  initial  boot  file      (specified  in  the  filename  statement) is to be loaded.  Server-name should be a numeric IP address or a domain      name.  If no next-server statement applies to a given client, the address 0.0.0.0 is used._

However in the lease file next-server is stored as an IP with four hex octets. I'd need to test this, maybe some changes are required on the smart proxy level.

Anyway, how you like my idea to ditch /serverName call which never made sense to me and introducing new Subnet field called "TFTP server name" which could be either IPv4 or a hostname.

File: lib/proxy_api/tftp.rb
Comment (ekohl) [inline lib/proxy_api/tftp.rb]: I actually dislike doing it here. This class is supposed to just model the proxy API and not do anything else.

File: app/models/concerns/orchestration/dhcp.rb
Comment (ekohl) [reply app/models/concerns/orchestration/dhcp.rb:91]: > However in the lease file next-server is stored as an IP with four hex octets. I'd need to test this, maybe some changes are required on the smart proxy level.

Then we'd leave it up to the DHCP server implementation. Looks like our omapi implementation doesn't care about an IP. You're right that our lease file parser could break on this.

> Anyway, how you like my idea to ditch /serverName call which never made sense to me and introducing new Subnet field called "TFTP server name" which could be either IPv4 or a hostname.

Since we introduced capabilities, we also expose the setting. This means you can do `smart_proxy.setting('tftp', 'tftp_servername')`. If it registered correctly, then you have the value in your local database. We actually don't make use of this now.

Moving it to the Subnet model might make sense since it's more flexible. On the other hand, the current model of having the proxy export it actually gives you service discovery. You expose a TFTP service and that service knows how it's reachable (since you configured it).

Comment (lzap) [reply app/models/concerns/orchestration/dhcp.rb:91]: > Since we introduced capabilities, we also expose the setting. This means you can do smart_proxy.setting('tftp', 'tftp_servername').

Sure but why this setting is on the smart proxy in the first place? Smart proxy don't need to know this, it's the Foreman what uses this bit of information when sending DHCP create request. I believe this setting should be completely removed from proxy. The only purpose of this setting is to be returned via the API, why?

```
[lzap@box smart-proxy]$ ag tftp_servername
config/settings.d/tftp.yml
5:#:tftp_servername: tftp.domain.com
9::tftp_servername: 192.168.100.1

config/settings.d/tftp.yml.example
7:#:tftp_servername: tftp.domain.com

modules/tftp/tftp_plugin.rb
13:    expose_setting :tftp_servername

modules/tftp/tftp_api.rb
70:       {"serverName" => (Proxy::TFTP::Plugin.settings.tftp_servername || "")}.to_json

test/tftp/integration_test.rb
15:    Proxy::DefaultModuleLoader.any_instance.expects(:load_configuration_file).with('tftp.yml').returns(enabled: true, tftproot: '/var/lib/tftpboot', tftp_servername: 'tftp.example.com')
26:    expected_settings = { 'tftp_servername' => 'tftp.example.com' }

test/tftp/tftp_api_test.rb
115:    Proxy::TFTP::Plugin.settings.stubs(:tftp_servername).returns("servername")

test/migrations/migration_settings.yml
18::tftp_servername: 192.168.122.1

test/migrations/monolithic_config_file_migration_test.rb
35:                 :tftp_servername => "192.168.122.1"
```

Comment (lzap) [reply app/models/concerns/orchestration/dhcp.rb:91]: Well, lemme ask https://projects.theforeman.org/issues/800 :-)

Comment (ekohl) [reply app/models/concerns/orchestration/dhcp.rb:91]: The purpose is service discovery. Let's say you have the following setup:
* foreman.example.com
* dhcp.example.com
* tftp.example.com

Now Foreman can query the foreman-proxy on tftp.example.com which address clients should use. Then on dhcp.example.com it can create the DHCP record with the correct TFTP server address.

If you didn't have this, then the user needs to manually do this. Since multiple DHCP servers can share the same TFTP proxy this is actually useful. All the user has to do is select the correct proxy for TFTP.

Comment (lzap) [reply app/models/concerns/orchestration/dhcp.rb:91]: Because if I am right, we can simply remove this completely, create Subnet.next-server DB field and warn user to either use IPv4 or hostname (but have correct DNS entry). And test this end to end if there are no changes required in our DHCP providers (I am little bit worried that some providers might expect IPv4 only). But that's it.

Comment (lzap) [reply app/models/concerns/orchestration/dhcp.rb:91]: You are right, let's simply improve the DNS resolution in this PR.

**Comment by lzap** [comment]
@ares we came to conclusion that it's best to only improve the DNS resolution and refactor the code to move it to better place (probably Subnet).

**Comment by lzap** [comment]
Observations from todays testing:

* ISC DHCP module does not allow next-server option to be hostname (omshell silently ignores the next-server statement, it does not appear in leases file at all)
* DHCP API allows hostnames and IP addresses
* our ISC DHCP module does actually perform reverse DNS search in case hostname is provided

It looks to me that we should remove this from Foreman and simply allow hostnames to be set. Individual proxy DHCP modules must make sure they are converted to IPv4 when needed.

**Comment by lzap** [comment]
CORRECTION:

* ISC DHCP module does allow next-server option to be hostname if it's properly escaped for omshell
* DHCP API allows hostnames and IP addresses
* our ISC DHCP module does actually perform reverse DNS search in case hostname is provided
* and if that PTR resolving fails, it will use the hostname (and ISC DHCP will accept it)

We don't need to have this in core at all. Let's just let the plugin authors and users know that this is changing.

**Comment by lzap** [comment]
ADDITIONAL FIND:

The PTR reverse code was not working correctly in proxy, I am gonna fix this in https://github.com/theforeman/smart-proxy/pull/657 and add tests.

WOW. That's nasty.

File: app/models/concerns/orchestration/dhcp.rb
Comment (ekohl) [inline app/models/concerns/orchestration/dhcp.rb:91]: This is the only use of `NicIpResolver`. Let's drop it if we no longer use it.

**Review by ekohl — CHANGES_REQUESTED** [review]
Currently this requires a fix that's hidden in https://github.com/theforeman/smart-proxy/pull/657 but I'd prefer that it gets its own Redmine issue for visibility.

It requires at least a smart-proxy that accepts a DNS name. I'm wondering if we should expose a capability in the module that lets us detect this automatically and fall back to the old behavior if it doesn't. That should make upgrading safer.

cc @tbrisker

File: app/models/concerns/orchestration/dhcp.rb
Comment (lzap) [reply app/models/concerns/orchestration/dhcp.rb:91]: Will gladly drop that.

**Comment by lzap** [comment]
> It requires at least a smart-proxy that accepts a DNS name. I'm wondering if we should expose a capability in the module that lets us detect this automatically and fall back to the old behavior if it doesn't. That should make upgrading safer.

While I generally appreciate good upgrade experience, this is actually a bug in smart proxy which we are fixing in the next release. Any workaround will lead to keeping the code in core, I would like to get rid of it.

How about this - let's delete unwanted code (my PR) and write a Upgrade Note about it. If it cease to work, the workaround is super easy - just set tftp-server-name option in proxy YAML (installer can do this). Or upgrade smart proxy.

**Comment by ekohl** [comment]
I'd be ok with a clear description. This might actually be an interesting blog as well with more depth. The current situation, the issues with it, path taken to debug and the new situation.

**Comment by lzap** [comment]
Irrelevant failure, it was actually green I accidentally pushed twice. Going to write a blogpost and attach link to the PR.

**Comment by lzap** [comment]
@amirfefer this test which you wrote randomly fails, can you take a look please?

```
Expected false to be truthy. (Minitest::Assertion)
/home/jenkins/workspace/test_develop_pr_core/database/postgresql/ruby/2.6/slave/fast/test/models/concerns/audit_extensions_test.rb:36    

  test "audited changes field can be greater then 65K bytes" do
    prov_template = templates(:mystring)
    prov_template.template = "0000000000" * 3500
    as_admin do
      assert prov_template.save!
      prov_template.template = "1111111111" * 3500
      assert prov_template.save!
    end
    assert Audit.last.audited_changes.to_s.bytesize > 66000
  end
```

**Comment by lzap** [comment]
Here is the long story. It's a wiki, feel free to correct my statements if I am not right. It's a bunch of things: https://community.theforeman.org/t/foreman-and-dhcp-options/15127

I felt like it's not the right time for the blogpost atm its bunch off opened PRs. I would really like to push this into 1.24.

**Review by ekohl — DISMISSED** [review]
I'd appreciate an ack from another maintainer since I feel I was very heavily involved in the current design. @ares or @tbrisker perhaps?

**Review by tbrisker — COMMENTED** [review]
I'm not too familiar with the whole tftp workflow, but if I understand correctly, the idea here is to stop trying to guess the next-server entry's ip in foreman and instead have the proxy handle this (so i guess we would already in that case get the ip when asking for `tftp.bootServer`)? Would the case of guessing the subnet's proxy hostname still be valid or should that one also be dropped?
I'm assuming this needs to be merged in accordance with the related proxy PR, right? and also an upgrade note please (if not done yet).

**Comment by amirfefer** [comment]
@lzap, I don't recall this test (2 years...) and I'm out of capacity, can we skip that test atm?

**Comment by lzap** [comment]
> Would the case of guessing the subnet's proxy hostname still be valid or should that one also be dropped?

Foreman still uses Smart proxy hostname in case TFTP next-server is not set explicitly by user, but it does not perform reverse DNS query anymore and passes the hostname around. When it gets to the DHCP smart proxy module it will do the reverse DNS query when needed - this is different per individual DHCP implementation.

The Smart proxy PR must be merged first, you are correct: https://github.com/theforeman/smart-proxy/pull/657

I will do Upgrade note, but no actions are required from the user, just in case they use some unofficial DHCP plugin that expects the TFTP next server to be always an IP address.

**Comment by lzap** [comment]
@ekohl @tbrisker hopefully I answered all concerns. I do not want to delay this any longer. I asked @ares to take a look but he can't find time. Let's merge this please, this is terribly broken and causes lot of pain for all users it can hardly be worse. ;-)

**Comment by lzap** [comment]
@tbrisker @ekohl ping - since we have a fresh release out, let's just merge this and see how it works in the wild?

**Comment by ekohl** [comment]
It isn't entirely clear to me if the current proxy accepts a DNS name. In https://github.com/theforeman/smart-proxy/pull/657 there is a hidden fix in the omapi provider but do we need that for this to work?

**Comment by lzap** [comment]
Proxy with the patch you referred does accept hostname and ISC DHCP module passes that into the server via omshell correctly, it ends up in the `dhcpd.leases` file. I have tested this end to end!

**Comment by ekohl** [comment]
But without the patch it doesn't? That would mean that users need to upgrade. There are two ways of dealing with this:
* Upgrade note: leave it to the user to make sure this works
* Capability

Advertising a capability on the proxy side would allow Foreman to know if it can use the modern version and send a name while keeping compatibility with the older version and send an IP. This would align with the goal to allow incremental upgrades rather than the big all-in-one upgrade.

**Comment by lzap** [comment]
Hmm, that does not make the code easier to understand but little bit more complex actually. Also do we refresh features during the upgrade? Until feature are refreshed, users might be stuck with the old behavior forever. I am not so keen into this, but if you insist...

**Comment by lzap** [comment]
The smart-proxy PR is now this one: https://github.com/theforeman/smart-proxy/pull/682

**Comment by ekohl** [comment]
You're right that it is more complicated, but downstream has a goal to make upgrades smoother and allow diverse versions to cooperate. I'd like @tbrisker's opinion but since he's on PTO maybe @ares could give a second opinion about the past 3 comments in this PR.

> Hmm, that does not make the code easier to understand but little bit more complex actually. Also do we refresh features during the upgrade? Until feature are refreshed, users might be stuck with the old behavior forever. I am not so keen into this, but if you insist...

Currently we may be relying on side effects in the installer. We also have the issue that capabilities are only currently only available over HTTPS which may be an issue for development setups. That's why I'm not insisting but I do want to make a well thought out choice here.

> The smart-proxy PR is now this one: [theforeman/smart-proxy#682](https://github.com/theforeman/smart-proxy/pull/682)

Thanks! I do think presenting the capability is a good thing regardless of whether we use it.

**Comment by lzap** [comment]
> Thanks! I do think presenting the capability is a good thing regardless of whether we use it.

How about an error telling the user to upgrade the smart proxy? But if you insist I can probably just deprecate the code if this makes you to merge this :-)

**Comment by ekohl** [comment]
> How about an error telling the user to upgrade the smart proxy?

I like that compromise! We don't have to drag old code but the user knows knows how to make it work.

**Comment by lzap** [comment]
@ares @ekohl added exception to be thrown, added an extra test for it.

**Comment by lzap** [comment]
I am going to add deprecation if capability is not present and keep the current behavior.

**Comment by lzap** [comment]
Rebased, the deprecated behaviour is back now with a warning and test for this. Let's remove this later! Thanks all.

File: app/models/concerns/orchestration/dhcp.rb
Comment (ekohl) [inline app/models/concerns/orchestration/dhcp.rb]: I'd prefer to capabilities to be exact matches. In this case we're looking for `dhcp_filename_hostname`.

Comment (lzap) [reply app/models/concerns/orchestration/dhcp.rb]: Makes sense, rebased.

File: app/models/concerns/orchestration/dhcp.rb
Comment (ekohl) [inline app/models/concerns/orchestration/dhcp.rb]: Isn't this equivalent to:
```suggestion
    if subnet.dhcp.capabilities(:DHCP).include?("dhcp_filename_hostname")
```

Not needed for this PR, but a helper on the smart proxy model would be nice:
```ruby
def has_capability?(feature, capability)
  caps = capabilities(feature)
  caps && caps.include?(capability)
end
```

Comment (lzap) [reply app/models/concerns/orchestration/dhcp.rb]: Added the helper.

**Comment by lzap** [comment]
Rebased.

File: app/models/smart_proxy.rb
Comment (ekohl) [reply app/models/smart_proxy.rb]: That's my bad for not suggesting that :)

Comment (lzap) [reply app/models/smart_proxy.rb]: Oh no I come after lunch for green color and this... ;-) Lemme fix that.

**Comment by lzap** [comment]
Rebased.

File: app/models/concerns/orchestration/dhcp.rb
Comment (ekohl) [inline app/models/concerns/orchestration/dhcp.rb:94]: Technically the whole indenting in else here is not needed since the `if` statement returns. That makes the diff easier to read.

**Review by ekohl — APPROVED** [review]
Overall :+1:

File: app/models/concerns/orchestration/dhcp.rb
Comment (lzap) [reply app/models/concerns/orchestration/dhcp.rb:94]: Well we are going to drop this in 6 months so probably not a big deal.

Comment (ekohl) [reply app/models/concerns/orchestration/dhcp.rb:94]: My reasoning is mostly that the diff is smaller so it's easier to see the actual differences vs some whitespace changes, but I won't block on it. Just a hint for the future.

**Comment by lzap** [comment]
It's green @ares or @ekohl

File: app/models/smart_proxy.rb
Comment (ares) [inline app/models/smart_proxy.rb:96]: nice

**Review by ares — APPROVED** [review]
Thanks @lzap, merging!

**Pull request merged**
